csharp/src/IceLocatorDiscovery/PluginI.cs

```diff
@@ -256,350 +173,224 @@ public void
-
-                    if (request != null)
+                    if (_findLocatorTask == null)
                     {
-                        _pendingRequests.Add(request);
-                    }
-
-                    if (!_pending) // No request in progress
-                    {
-                        _pending = true;
-                        _pendingRetryCount = _retryCount;
-                        _failureCount = 0;
-                        try
-                        {
-                            if (_traceLevel > 1)
-                            {
-                                var s = new StringBuilder("looking up locator:\nlookup = ");
-                                s.Append(_lookup);
-                                if (_instanceName.Length == 0)
-                                {
-                                    s.Append("\ninstance name = ").Append(_instanceName);
-                                }
-                                _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
-                            }
-
-                            foreach (KeyValuePair<ILookupPrx, ILookupReplyPrx?> l in _lookups)
-                            {
-                                _ = FindLocator(l.Key, l.Value!); // Send multicast request.
-                            }
-                            _timer.Schedule(this, _timeout);
-                        }
-                        catch (System.Exception ex)
-                        {
-                            if (_traceLevel > 0)
-                            {
-                                var s = new StringBuilder("locator lookup failed:\nlookup = ");
-                                s.Append(_lookup);
-                                if (_instanceName.Length == 0)
-                                {
-                                    s.Append("\ninstance name = ").Append(_instanceName);
-                                }
-                                s.Append("\n").Append(ex);
-                                _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
-                            }
-
-                            foreach (Request req in _pendingRequests)
-                            {
-                                _ = req.Invoke(_voidLocator);
-                            }
-                            _pendingRequests.Clear();
-                            _pendingRetryCount = 0;
-                            _pending = false;
-                        }
+                        _findLocatorTask = FindLocatorAsync();
                     }
+                    findLocatorTask = _findLocatorTask;
                 }
             }
+            return await findLocatorTask.ConfigureAwait(false);
         }
 
-        private void Exception(System.Exception ex)
+        private async Task<ILocatorPrx> FindLocatorAsync()
         {
-            lock (this)
+            TaskCompletionSource<ILocatorPrx> completionSource;
```

Comment:
This temp isn't really needed if FoundLocator doesn't clear the task completion source.

Reply:
Fixed